kolibri/core/analytics/utils.py

```diff
@@ -423,6 +423,9 @@ def perform_ping(started, server=DEFAULT_SERVER_URL):
 
     language = get_device_setting("language_id", "")
 
+    if not isinstance(started, datetime.datetime):
+        started = datetime.datetime.strptime(started)
+
```

Comment:
@rtibbles :100:.

Reply:
If all of the serialization/deserialization needs to retain typings for things that can't be serialized (so far just dates and errors), does this mean we should have some sort of type-decoding dictionary, and an extra, reserved attribute in the serialized table that lists the type of each part that isn't a primitive?

I saw people warning against checking each value to see if it's a datetime because that's slow and can, in rare cases, yield false positives. Following that advice, there would be only two ways to do it; first, checking in each function that is expecting a datetime (or error); second, storing more data in the object, and using that data to decode any objects that were turned into just strings.

I'm kinda liking the idea of going back and doing the second one, it's just a bit harder initially.

Reply:
This is one reason why moving more of the Job data into separate ORMJob columns might be advantageous, as it would reduce the overhead for this kind of type checking.

If we had a top level `exception` column on the ORMJob table (that is nullable) then we don't need a general purpose parser to handle serialized `Exception` objects.

I think the main remaining issue then is that we use a `datetime` object as a function argument and we are expecting to be able to serialize/deserialize that. I think we could use a try/except to handle deserialization, using the Django utils parse datetime.

The Python JSONDecoder object allows passing `object_hooks` which try to do further deserialization, so, for example, we would try to coerce any strings to datetimes using the Django parse function, and just return the original string if it failed.

See the documentation here: https://docs.python.org/3/library/json.html#json.JSONDecoder and an example here: https://stackoverflow.com/a/10734224 (note they are using a different parsing function).

Reply:
@rtibbles I have a different opinion on this one.

I believe the task function should take care of parsing the datetime string. The reason being we can have in rare cases some false positives. I would say, we must establish that the arguments passed to the task function will all be valid-JSON types ONLY. So, then, it will be the job of the task function to take care of parsing the datetime string. 

This will be more performant and the task function will have 100% clarity on what's going on.

Reply:
In that case, we should not use `DjangoJSONEncoder` and just use regular `json.dumps`, so as not to have any ambiguity. That means that the function initializing the pingback task would have to pre-serialize the argument.

I am fine with this if we want to be strict.